---
title: "Find and annotate DNA mutation clusters with cMut"
author: "Alex Janse"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(cMut)
```
<html>
<head>

<style type="text/css">
abbr {    
    border-bottom: 1px dotted #000;
    text-decoration: none;
}
</style>
</head>

</html>

The cMut (clustered mutation) package provide functions to find and annotate clustered mutations.
In the following examples, the functions will be explained and used.

# Context {#Context}
* [Purpose of cMut](#Purpose)
* [Create Random Mutations](#Create)
* [Default Mutation Patterns](#Default)
* [Match Mutations to Mutation Patterns](#Match)
* [Identify and Annotate Clusters](#Identify)
* [Group Clusters and Find Cluster Patterns](#Group)
* [Summarize the Found Patterns](#Summarize)
* [Determine the Enrichment](#Determine)
* [Interpretation of Results](#Interpretation)

\pagebreak

# Purpose of cMut {#Purpose}
When learning about DNA, we've learned that all the cells in our body (that contain DNA) have the same DNA throughout the body. Although that is mostly true, when we get older the more variations arise in different cells due exposure to different mutators, error-prone replication and different speed of decay of the telomeres.  
The variations are the best displayed when looking at the mutations in DNA sequence (sequence of nucleotides A, C, G and T in different orders and frequencies). These mutations can exist in many shapes and sizes; from single nucleotide variant (SNV) till large copy number variants (CNV).  
In this package we're interested in finding and annotating clustered SNV (cSNV a.k.a. multi-nucleotide mutation; MNM). We also try to link them with known mutator patterns and check if the enrichment is valid.  
  
Therefore the following questions arise:  
  
* **Bioinformatics:**  
  Is there an enrichment of clustered mutations that match with certain mutation patterns?
* **Biology:**  
  Which mutators are responsible for the clustered mutations?
  
With this package we hope to answer the bioinformatics question and therefore give a clue about the answer for the biology question.  
  
~[Back to Context](#Context)~  

\pagebreak

# Create Random Mutations {#Create}
The `createRandomMutation()` function will create random single nucleotide variant (SNV) mutations in the human reference genome GRCh37/hg19 (or GRch38/hg38 if `refGenomeHg19 = FALSE`) and store them in a tibble (data.frame like object from the tidyverse package).
```{r echo = TRUE, results = 'hide'}
randomData <- createRandomMutations(nMut = 10)
randomData
```
```{r echo=FALSE}
knitr::kable(randomData,align = 'c')
```
  
For more information about the table, use:
```{r include = TRUE}
cat(comment(randomData))
```
  
~[Back to Context](#Context)~  
  
\pagebreak

# Default Mutation Patterns {#Default}
The following functions below use a search table to see if it is possible to match the mutation with certain mutation patterns. If no search table is sent through the <abbr title = 'The objects you sent along when calling a function like in "fun(x = 1)" , x is the argument.'>arguments</abbr>, a default table `mutationPatterns` by the package author is used.

In this table, there are two kinds of patterns:  
  
### Mutation Patterns
```{r echo = FALSE}
knitr::kable(getSearchPatterns(F,F,F)[1:9,], algin = 'c')
```  
  
Patterns with 1 reference, 1 alternative, with surrounding nucleotides and a maximum distance with a number or NA. These are used in the [identifyAndAnnotateClusters()](#Identify) and [linkPatterns()](#Match) functions. These patterns will be called 'mutation patterns' in the documentation of this package.

### Cluster Patterns
```{r echo = FALSE}
knitr::kable(getSearchPatterns(F,F,F)[10:15,], algin = 'c')
``` 
  
Patterns with more than 1 or no reference, more than 1 alternative, no surrounding nucleotides and always a maximum distance (in this case: `PolZeta` and `PolZeta.endOnly`). These are used in the [groupClusters()](#Group) function. These are separately evaluated from the mutation patterns because these patterns depend on the order and distance of the mutations within a cluster. These patterns will be called 'cluster patterns' in the documentation of this package.  
  
## Nucleotide symbols
As seen in the table there are nucleotides that do not match with the well known A,G,C,T nucleotides. That is because these are not nucleotides but symbols that represent the nucleotides as shown in the `dnaAlphabet` table:
```{r echo = TRUE, results='hide'}
dnaAlphabet
```
```{r echo=FALSE}
knitr::kable(dnaAlphabet,align = 'c')
```
  
For more information, see the <abbr title = 'International Union of Pure and Applied Chemistry'>IUPAC</abbr> annotations section of [this Wikipedia page](<https://en.wikipedia.org/wiki/Nucleic_acid_notation>).  
  
~[Back to Context](#Context)~  
  
\pagebreak

# Match Mutations to Mutation Patterns {#Match}
The `linkPatterns()` function will check if the provided mutation can be matched with the so called 'mutation patterns' from the search pattern table and returns the ID's of the match.
The default search table is used if no search pattern table is sent with the `searchPattern` argument.  
See [Default mutation patterns](#Default) paragraph above for more information.  
  
A proper way of using this function with the default search table:
```{r include = TRUE}
linkPatterns(ref     = "C",
             alt     = "G",
             context = "CT.AT")
```
  
If distance is important, use the `distance` argument to tell the distance to the nearest mutation within a cluster. Make sure that the search pattern table has a column with the maximum distance for the specific pattern.
```{r include = T}
linkPatterns(ref = "A", alt = "G", context = "TA.TA", distance = 550)
linkPatterns(ref = "A", alt = "G", context = "TA.TA", distance = 50)
```

As seen in the [Default mutation patterns](#Default) paragraph, it is possible to use alternative nucleotide symbols. However it is not possible to use these symbols in the look up mutation:
```{r include = TRUE, error = TRUE}
linkPatterns(ref = "C", alt = "D", context = "A.A")
```

By default the `linkPatterns()` function will also look at the <abbr title = "The sequence on the opposing strand in reverse so DNA is mostly written down as something like 5'AGCGT3' so the complement would be 3'TCGCA5' and then the reverse is 5'ACGCT3'. The reverse complement is good to know because most mutators are not strand specific and also checking the reverse complement you will look at both strands.">reverse complement</abbr> of each pattern in the sent search table:
```{r include = TRUE}
linkPatterns(ref     = "C",
             alt     = "G",
             context = "CT.AT")

linkPatterns(ref     = "G",
             alt     = "C",
             context = "AT.AG")
```
There are a few arguments that involves in searching for the reverse complement:
```{r include = TRUE}
# Example search pattern table
exampleSearch <- data.frame(stringsAsFactors = FALSE,
                            process          = c("id1", "id2"),
                            ref              = c("A","T"),
                            alt              = c("S","S"),
                            surrounding      = c("T.T","A."))
exampleSearch
```
The `searchReverseComplement` argument tells if the search pattern table as needs to be used in the reverse complement version
```{r include = TRUE}
linkPatterns(ref                     = "A",
             alt                     = "G",
             context                 = "GT.TC",
             searchPatterns          = exampleSearch,
             searchReverseComplement = TRUE)

linkPatterns(ref                     = "A",
             alt                     = "G",
             context                 = "GT.TC",
             searchPatterns          = exampleSearch,
             searchReverseComplement = FALSE)
```
The `reverseComplement` argument tells if the sent mutation (in this case A>G; GT.TC) needs to be converted to the reverse complement version of itself. Make sure that searchReverseComplement is FALSE otherwise it shall give the same results as above.
```{r include = TRUE}
linkPatterns(ref                     = "A",
             alt                     = "G",
             context                 = "GT.TC",
             searchPatterns          = exampleSearch,
             searchReverseComplement = FALSE,
             reverseComplement       = TRUE)

linkPatterns(ref                     = "A",
             alt                     = "G",
             context                 = "GT.TC",
             searchPatterns          = exampleSearch,
             searchReverseComplement = FALSE,
             reverseComplement       = FALSE)
```
The last argument is `renameReverse`. This arguments tells if the IDs of the reverse complement version of search pattern table needs to be called differently: \*ID\* [Rev.Com.]. This is nice if you want to know if the match came from the reverse complement or not.
```{r include = TRUE}
linkPatterns(ref                     = "A",
             alt                     = "G",
             context                 = "GT.TC",
             searchPatterns          = exampleSearch,
             searchReverseComplement = TRUE,
             renameReverse           = TRUE)

linkPatterns(ref                     = "A",
             alt                     = "G",
             context                 = "GT.TC",
             searchPatterns          = exampleSearch,
             searchReverseComplement = TRUE,
             renameReverse           = FALSE)
```
Besides the arguments `ref`, `alt` and `context`, all arguments for `linkPatterns()` function are also available in the [identifyAndAnnotateClusters()](#Identify) and [groupClusters()](#Group) functions.  
  
~[Back to Context](#Context)~  
  
\pagebreak

# Identify and Annotate Clusters {#Identify}
The `identifyAndAnnotateClusters()` function searches for clusters by looking per sample per chromosome if there are mutations within the range of the maximum distance. It also applies [linkPatterns()](#Match) to every clustered mutation row of the sent table. Beside those functions, it will also add the distance towards the nearest mutations and columns with <abbr title = "Datatype that is either TRUE or FALSE">Booleans</abbr> whether or not the function it referring to gave results.  
```{r echo = TRUE, results = 'hide'}
results <- identifyAndAnnotateClusters(dataTable   = testDataSet, 
                                       maxDistance = 20000,
                                       asTibble    = FALSE)
head(results[results$is.linked, ])
```
```{r echo=FALSE}
knitr::kable(head(results[results$is.linked, ]), align = 'c')
```
  
Please make sure that the <abbr title = 'The values that are already filled in for the function's arguments. See the usage section at the function's help page to see these values.'>default arguments</abbr> for column names (`chromHeader`, `sampleIdHeader`, `positionHeader`, `refHeader`, `altHeader` and `contextHeader`) correspond with the column headers of the sent table or change it if needed. The presence of other columns should not be a problem but make sure that there are no duplicate column names that are used in the arguments.  For an example of the input for the `dataTable` argument see the `testDataSet` table or use the [createRandomMutations()](#Create) function.   
  
When using your own search pattern table with column names or others arguments that do not match with the default, please make sure to change the corresponding `search...` arguments.
```{r echo = TRUE, results = "hide"}
# Example of using you own search pattern table:

# Build an example table
testPatterns <- data.frame(id          = c("testPat1", "testPat2"),
                           reference   = c("A", "G"),
                           alternative = c("T", "C"), 
                           dist        = c(NA, NA),
                           sur         = c("-", "DB-Y"))
testPatterns
```
```{r echo = FALSE}
knitr::kable(testPatterns, align = 'c')
```
```{r echo = TRUE, results = "hide"}
# Using the table on the function:
results2 <- identifyAndAnnotateClusters(dataTable              = testDataSet, 
                                       maxDistance             = 20000,
                                       asTibble                = FALSE,
                                       linkPatterns            = TRUE,
                                       searchPatterns          = testPatterns,
                                       searchIdHeader          = "id",
                                       searchRefHeader         = "reference",
                                       searchAltHeader         = "alternative",
                                       searchDistanceHeader    = "dist",
                                       searchContextHeader     = "sur",
                                       searchMutationSymbol    = "-",
                                       searchReverseComplement = FALSE)
tail(results2[results2$is.linked,])
```
```{r echo = FALSE}
knitr::kable(tail(results2[results2$is.linked,]), align = 'c')
```
  
For more information about the added columns, use:
```{r include = TRUE}
cat(comment(results))
```
  
~[Back to Context](#Context)~  

\pagebreak

# Group Clusters and Find Cluster Patterns {#Group}
The `groupClusters()` function will group the data based on the cluster IDs and shows a summary of each cluster. It also adds a column with the intersection (overlap) of the patterns per cluster and found cluster patterns (See the [Default mutation patterns](#Default) chapter above for more information about cluster patterns). The `groupClusters()` function is best used as a followup function for [identifyAndAnnotateClusters()](#Identify).  
```{r echo = TRUE, results = 'hide'}
results <- identifyAndAnnotateClusters(dataTable    = testDataSet, 
                                       maxDistance  = 20000)

groups <- groupClusters(dataTable = results)

head(groups[groups$has.clusterPatterns | 
              groups$has.intersect, -which(names(groups) == "cMuts")])
```
```{r echo=FALSE}
knitr::kable(groups[groups$has.clusterPatterns | groups$has.intersect, -which(names(groups) == "cMuts")], align =  'c')
```
  
The same arguments that have been used in the [linkPatterns()](#Match) function about reverse complement are also present in the `groupClusters()` function and are relevant when `searchClusterPatterns = TRUE`. See the [linkPatterns()](#Match) paragraph for more information about what each argument does.  
  
Please note that if `renameReverse = TRUE` is used when using the [identifyAndAnnotateClusters()](#Identify) function the `groupClusters()` function will find no intersection if the cluster contains e.g. row 1 = "MMR" and row 2 = "MMR [Rev.Com.]". This can come in handy if you wish to find patterns located on the same strand but still want to search in both the minus and the plus strand.
```{r include = TRUE}
results2 <- identifyAndAnnotateClusters(dataTable    = testDataSet, 
                                       maxDistance   = 20000,
                                       renameReverse = TRUE)

groups2 <- groupClusters(dataTable     = results,
                         renameReverse = TRUE)

groups2[groups2$has.clusterPatterns | groups2$has.intersect,"foundPatterns"][[1]]
```
  
For more information about the columns of the results, use:
```{r include = TRUE}
cat(comment(groups))
```
  
~[Back to Context](#Context)~  
  
\pagebreak

# Summarize the Found Patterns {#Summarize}
To summarize the clusters that are given out by [groupClusters()](#Group), the function getSummaryPatterns() gives the number of clustered mutations affected by each mutation pattern. The `unidentified` row contains the information about the clusters that didn't have a linked pattern.
```{r echo = TRUE, results='hide'}
results <- identifyAndAnnotateClusters(dataTable    = testDataSet, 
                                       maxDistance  = 20000)

groups <- groupClusters(dataTable = results)

getSummaryPatterns(groupedClusters = groups) 
```
```{r echo = FALSE}
knitr::kable(getSummaryPatterns(groups), align = 'c')
```
  
If the `groupedClusters` table contains patterns that are not present in the `searchPatterns` table, then it will be marked as unidentified together with clusters without patterns.
```{r echo = TRUE, results = 'hide'}
example <- getSearchPatterns()[1, ] # Default search table with only the first row
example
```
```{r echo = FALSE}
knitr::kable(example, align = 'c')
```
```{r echo = TRUE, results = 'hide'}
getSummaryPatterns(groupedClusters = groups,
                   searchPatterns  = example)
```
```{r echo = FALSE}
knitr::kable(getSummaryPatterns(groups, searchPatterns = example), align = 'c')
```
  
Please note that if you used `renameReverse = TRUE` in [identifyAndAnnotateClusters()](#Identify) and/or [groupClusters()](#Group) functions than also use it in the `getSummaryPatterns()` function. Otherwise the reverse complement patterns will end up in the "Unidentified" row.
```{r echo = TRUE, results='hide'}
getSummaryPatterns(groupedClusters = groups2,
                   renameReverse   = TRUE)
```
```{r echo = FALSE}
knitr::kable(getSummaryPatterns(groups2,
                   renameReverse  = TRUE), align = 'c')
```
```{r echo = TRUE, results = 'hide'}
getSummaryPatterns(groupedClusters = groups2)
```
```{r echo = FALSE}
knitr::kable(getSummaryPatterns(groups2), align = 'c')
```
  
~[Back to Context](#Context)~  
  
\pagebreak

# Determine the Enrichment {#Determine}
In order to tell if the found results from the [getSummaryPatterns()](#Summarize) function are truly enriched or happen by chance alone, a shuffle function is needed to validate the enrichment. The `shuffleMutations()` is able to shuffle the reference, alternative and surrounding nucleotides at random. Then it execute the [identifyAndAnnotateClusters()](#Identify) and [groupCluster()](#Group) function. Finally it will keep track on how many times the patterns are found which might give a hint of reliability of the results of the real data. With the `nBootstrap` argument it is possible to choose the number of repeats this function has to make. Make sure to use only the mutations that are clustered in the real data for a proper indication.  
  
To save time, you can use the `no.cores` argument to enable parallelization. It takes the number of cores that are used for the computations. The default is the maximum number of cores available to your system.
```{r echo = TRUE, results = 'hide', message=FALSE}
clusteredMut <- results[results$is.clustered == TRUE,]
shuffle <- shuffleMutations(dataTable  = clusteredMut,
                            nBootstrap = 5,
                            no.cores   = 1)
# 1 core is used because of the limitation of CRAN check system
shuffle
```
```{r echo = FALSE}
# The hidden while loops are to make sure that a good example is shown
while(shuffle[8,"percentage"][[1]] == 100){
  invisible(capture.output(shuffle <- shuffleMutations(dataTable  = clusteredMut,
                                                       nBootstrap = 5,
                                                       no.cores   = 1)))  
}
knitr::kable(shuffle)
```
  
For more information about the results:  
```{r include = TRUE}
cat(comment(shuffle))
```
  
If wanted it is also possible to get the results per bootstrap:  
```{r echo = TRUE, results='hide', message=FALSE}
clusteredMut <- results[results$is.clustered == TRUE, ]
shuffle <- shuffleMutations(dataTable             = clusteredMut,
                            nBootstrap            = 2,
                            no.cores              = 1,
                            returnEachBootstrap   = TRUE)
shuffle
```
```{r echo = FALSE}
knitr::kable(shuffle)
```
The row with "total" as ID, represent the number of mutation rows that were clustered. Can be used to calculate the percentages as shown in the result table when `returnEachBootstrap = FALSE`.  
  
  
~[Back to Context](#Context)~  
  
# Interpretation of Results {#Interpretation}
The following code will shows and example of how to use all the main function properly and how to interpret the results.
```{R eval=FALSE}
# Example data to work with:
data <- testDataSet

# Identify and Annotate clusters:
resultsPerMutation <- identifyAndAnnotateClusters(dataTable   = data,
                                                  maxDistance = 20000)
# Group the clusters
resultsPerCluster <- groupClusters(dataTable = resultsPerMutation)

# Summarize the pattern results
summary <- getSummaryPatterns(groupedClusters = resultsPerCluster)

# Validate the enrichment:

# Best to only sent clustered mutations to the shuffle function
shuffleData <- resultsPerMutation[resultsPerMutation$is.clustered, ]
# Use the shuffle function.
# We used the returnEachBootstrap argument so we'll be able to draw a graph from the results
shuffleResults <- shuffleMutations(dataTable           = shuffleData, 
                                   maxDistance         = 20000,
                                   nBootstrap          = 1000,
                                   returnEachBootstrap = TRUE)
```
After using these functions we can build a plot with the shuffle results and compare it with the found results from the testDataSet. Because building a plot can be done in many different ways we'll leave that up to you. 
In the following plots, the black bars represent the distribution of the randomness of finding a pattern with that corresponding percentage (x-axis). The red vertical line represent the percentage that we found in the testDataset.  
In the following image you can see an example when a pattern is enriched:
  
![](ExamplePlotEnriched.png)
As you can see, the red line is far outside the contribution of randomness and therefore indicate that the pattern is enriched.  
  
And in the following plot you see an example when it's not enriched:
  
![](ExamplePlotNotEnriched.png)  
In the plot you can see that the red line is within the contribution of randomness an is therefore not enriched.
  
Of course there are many other ways aswell to interpreted the results. For example you can look at the distances between mutations that share the same pattern or look in which chromosome the pattern is the most frequent. We'll leave this interpretation up to you but hopefully this package makes the analysis a bit easier.
  
### Bugs, questions, suggestions?
Feel free to the contact the creator! (^__^)  
alex.janse@radboudumc.nl
