<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Alex Janse" />

<meta name="date" content="2018-09-18" />

<title>Find and annotate mutational clusters.</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Find and annotate mutational clusters.</h1>
<h4 class="author"><em>Alex Janse</em></h4>
<h4 class="date"><em>2018-09-18</em></h4>



<p>The cMut package provide functions to find and annotate clustered mutations. In the following examples, the functions will be explained and used.</p>
<div id="createrandommutations" class="section level2">
<h2>createRandomMutations</h2>
<p>The <code>cMut::createRandomMutation()</code> function will create random single nucleotide variant (SNV) mutations in the human reference genome (GRCh37/hg19) and store them in a tibble (table object from the tidyverse package).</p>
<p>When determine the surrounding nucleotides, the function will take into account that there might be mutations in the table that correspond with the surrounding locations. In this case the mutated nucleotide will be added to the surrounding instead of the reference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">randomData &lt;-<span class="st"> </span>cMut<span class="op">::</span><span class="kw">createRandomMutations</span>(<span class="dt">nMut =</span> <span class="dv">1000</span>)
randomData
<span class="co">#&gt; # A tibble: 1,000 x 7</span>
<span class="co">#&gt;    chrom     start      stop ref   alt   sampleIDs  surrounding</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      &lt;chr&gt;      </span>
<span class="co">#&gt;  1 chr8   12927902  12927902 G     T     testSample TG.AG      </span>
<span class="co">#&gt;  2 chr14  23885661  23885661 A     T     testSample AG.GA      </span>
<span class="co">#&gt;  3 chr13  63288397  63288397 A     C     testSample AC.AA      </span>
<span class="co">#&gt;  4 chr2   54663337  54663337 G     A     testSample AA.AA      </span>
<span class="co">#&gt;  5 chr7  129426308 129426308 T     G     testSample GG.TT      </span>
<span class="co">#&gt;  6 chr5   18236170  18236170 G     A     testSample GA.TG      </span>
<span class="co">#&gt;  7 chrX   27246983  27246983 A     C     testSample CA.AT      </span>
<span class="co">#&gt;  8 chr14  77287293  77287293 T     C     testSample CA.GA      </span>
<span class="co">#&gt;  9 chr10  18472221  18472221 A     T     testSample GT.AC      </span>
<span class="co">#&gt; 10 chr2  183014749 183014749 A     T     testSample AA.GG      </span>
<span class="co">#&gt; # ... with 990 more rows</span></code></pre></div>
<p>If a tibble is inconvenience, use the <code>tibble = FALSE</code> parameter to get a data frame.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cMut<span class="op">::</span><span class="kw">createRandomMutations</span>(<span class="dt">nMut =</span> <span class="dv">10</span>, <span class="dt">tibble =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt;    chrom     start      stop ref alt  sampleIDs surrounding</span>
<span class="co">#&gt; 1   chr7  99679889  99679889   G   C testSample       GT.TA</span>
<span class="co">#&gt; 2   chr7 147665962 147665962   G   A testSample       TG.GC</span>
<span class="co">#&gt; 3  chr16   2327955   2327955   G   T testSample       CC.CA</span>
<span class="co">#&gt; 4   chrX  80593817  80593817   G   T testSample       CT.GC</span>
<span class="co">#&gt; 5   chr3 194280109 194280109   G   T testSample       GT.TA</span>
<span class="co">#&gt; 6   chr2  41735573  41735573   T   C testSample       TA.TG</span>
<span class="co">#&gt; 7   chr5  57381196  57381196   A   C testSample       AG.AA</span>
<span class="co">#&gt; 8  chr10 119039850 119039850   A   T testSample       AA.AA</span>
<span class="co">#&gt; 9   chrX 106186940 106186940   T   C testSample       AT.CT</span>
<span class="co">#&gt; 10  chr8  44593110  44593110   N   G testSample       NN.NN</span></code></pre></div>
</div>
<div id="identifyandannotateclusters" class="section level2">
<h2>identifyAndAnnotateClusters</h2>
<p>The <code>cMut::identifyAndAnnotateClusters()</code> applies the <code>cMut::identifyClusters()</code> and if wanted the <code>cMut::linkPatterns()</code> to every row of the sent table. Beside those functions, it will also add the distance towards the nearest mutations and columns with booleans whether or not the function it referring to gave results. Make sure that the default parameters for column names correspond with the column headers of the sent table and change if needed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">results &lt;-<span class="st"> </span>cMut<span class="op">::</span><span class="kw">identifyAndAnnotateClusters</span>(<span class="dt">x =</span> randomData, 
                                             <span class="dt">maxDistance =</span> <span class="dv">20000</span>,
                                             <span class="dt">tibble =</span> <span class="ot">FALSE</span>)
results[results<span class="op">$</span>is.clustered <span class="op">==</span><span class="st"> </span>T,]
<span class="co">#&gt;     chrom     start      stop ref alt  sampleIDs surrounding</span>
<span class="co">#&gt; 83  chr10   8651049   8651049   T   C testSample       CA.TC</span>
<span class="co">#&gt; 84  chr10   8655796   8655796   G   A testSample       TA.TC</span>
<span class="co">#&gt; 322 chr17  39148935  39148935   C   A testSample       CT.TT</span>
<span class="co">#&gt; 323 chr17  39157038  39157038   G   C testSample       AT.CA</span>
<span class="co">#&gt; 401  chr2  63230363  63230363   A   G testSample       CA.TG</span>
<span class="co">#&gt; 402  chr2  63247795  63247795   T   C testSample       TG.TT</span>
<span class="co">#&gt; 712  chr5 171442753 171442753   T   A testSample       CT.GG</span>
<span class="co">#&gt; 713  chr5 171444816 171444816   T   A testSample       TG.AG</span>
<span class="co">#&gt; 772  chr7  43429898  43429898   G   T testSample       GG.AG</span>
<span class="co">#&gt; 773  chr7  43445104  43445104   G   T testSample       CA.CG</span>
<span class="co">#&gt; 917  chr9 126792383 126792383   G   A testSample       CA.CA</span>
<span class="co">#&gt; 918  chr9 126806597 126806597   A   T testSample       AT.CA</span>
<span class="co">#&gt;               clusterId is.clustered distance</span>
<span class="co">#&gt; 83   chr10 testSample 6         TRUE     4746</span>
<span class="co">#&gt; 84   chr10 testSample 6         TRUE     4746</span>
<span class="co">#&gt; 322 chr17 testSample 12         TRUE     8102</span>
<span class="co">#&gt; 323 chr17 testSample 12         TRUE     8102</span>
<span class="co">#&gt; 401  chr2 testSample 27         TRUE    17431</span>
<span class="co">#&gt; 402  chr2 testSample 27         TRUE    17431</span>
<span class="co">#&gt; 712  chr5 testSample 55         TRUE     2062</span>
<span class="co">#&gt; 713  chr5 testSample 55         TRUE     2062</span>
<span class="co">#&gt; 772  chr7 testSample 13         TRUE    15205</span>
<span class="co">#&gt; 773  chr7 testSample 13         TRUE    15205</span>
<span class="co">#&gt; 917  chr9 testSample 48         TRUE    14213</span>
<span class="co">#&gt; 918  chr9 testSample 48         TRUE    14213</span></code></pre></div>
<p>If the <code>cMut::linkPatterns()</code> is needed, use the <code>linkPatterns = TRUE</code> parameter. See the <code>cMut::linkPatterns()</code> paragraph for more information</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">results &lt;-<span class="st"> </span>cMut<span class="op">::</span><span class="kw">identifyAndAnnotateClusters</span>(<span class="dt">x =</span> randomData, 
                                             <span class="dt">maxDistance =</span> <span class="dv">20000</span>,
                                             <span class="dt">tibble =</span> <span class="ot">FALSE</span>,
                                             <span class="dt">linkPatterns =</span> <span class="ot">TRUE</span>)
results[results<span class="op">$</span>is.linked <span class="op">==</span><span class="st"> </span>T,]
<span class="co">#&gt;     chrom     start      stop ref alt  sampleIDs surrounding</span>
<span class="co">#&gt; 83  chr10   8651049   8651049   T   C testSample       CA.TC</span>
<span class="co">#&gt; 322 chr17  39148935  39148935   C   A testSample       CT.TT</span>
<span class="co">#&gt; 323 chr17  39157038  39157038   G   C testSample       AT.CA</span>
<span class="co">#&gt; 401  chr2  63230363  63230363   A   G testSample       CA.TG</span>
<span class="co">#&gt; 402  chr2  63247795  63247795   T   C testSample       TG.TT</span>
<span class="co">#&gt; 713  chr5 171444816 171444816   T   A testSample       TG.AG</span>
<span class="co">#&gt; 772  chr7  43429898  43429898   G   T testSample       GG.AG</span>
<span class="co">#&gt; 917  chr9 126792383 126792383   G   A testSample       CA.CA</span>
<span class="co">#&gt; 918  chr9 126806597 126806597   A   T testSample       AT.CA</span>
<span class="co">#&gt;               clusterId is.clustered distance   linkedPatterns is.linked</span>
<span class="co">#&gt; 83   chr10 testSample 6         TRUE     4746              MMR      TRUE</span>
<span class="co">#&gt; 322 chr17 testSample 12         TRUE     8102           A3G/A1      TRUE</span>
<span class="co">#&gt; 323 chr17 testSample 12         TRUE     8102    AID[Rev.Com.]      TRUE</span>
<span class="co">#&gt; 401  chr2 testSample 27         TRUE    17431              MMR      TRUE</span>
<span class="co">#&gt; 402  chr2 testSample 27         TRUE    17431              MMR      TRUE</span>
<span class="co">#&gt; 713  chr5 testSample 55         TRUE     2062              MMR      TRUE</span>
<span class="co">#&gt; 772  chr7 testSample 13         TRUE    15205 A3G/A1[Rev.Com.]      TRUE</span>
<span class="co">#&gt; 917  chr9 testSample 48         TRUE    14213    AID[Rev.Com.]      TRUE</span>
<span class="co">#&gt; 918  chr9 testSample 48         TRUE    14213              MMR      TRUE</span></code></pre></div>
</div>
<div id="identifyclusters.r" class="section level2">
<h2>identifyClusters.R</h2>
<p>The <code>cMut::identifyCluster()</code> function finds clusters in a table and returns a vector with cluster IDs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">resultVector &lt;-<span class="st"> </span>cMut<span class="op">::</span><span class="kw">identifyClusters</span>(<span class="dt">x =</span> randomData,
                 <span class="dt">maxDistance =</span> <span class="dv">20000</span>,
                 <span class="dt">chromHeader =</span> <span class="st">&quot;chrom&quot;</span>,
                 <span class="dt">positionHeader =</span> <span class="st">&quot;start&quot;</span>,
                 <span class="dt">sampleIdHeader =</span> <span class="st">&quot;sampleIDs&quot;</span>)
<span class="kw">head</span>(resultVector)
<span class="co">#&gt; [1] &quot;chr8 testSample 1&quot; &quot;chr8 testSample 2&quot; &quot;chr8 testSample 3&quot;</span>
<span class="co">#&gt; [4] &quot;&quot;                  &quot;&quot;                  &quot;&quot;</span></code></pre></div>
</div>
<div id="linkpatterns" class="section level2">
<h2>linkPatterns</h2>
<p>The <code>cMut::linkPatterns()</code> function can check if the provided mutation can be matched with known mutation patterns in a table and returns the ID of the match. If no table is sent the default table is used:</p>
<pre><code>#&gt; [1] &quot;DEFAULT TABLE:&quot;
#&gt; # A tibble: 9 x 4
#&gt;   proces            ref   alt   surrounding
#&gt;   &lt;chr&gt;             &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      
#&gt; 1 AID               C     D     WR.        
#&gt; 2 AID(hotspot)      C     D     WR.Y       
#&gt; 3 AID(pref.hotspot) C     D     AG.T       
#&gt; 4 MMR               A     B     W.         
#&gt; 5 MMR               T     V     .W         
#&gt; 6 A3G/A1            C     D     T.         
#&gt; 7 A3F               C     D     C.         
#&gt; 8 A3A               C     T     YT.A       
#&gt; 9 A3B               C     T     RT.A</code></pre>
<p>A proper way of using this function with the default table:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cMut<span class="op">::</span><span class="kw">linkPatterns</span>(<span class="st">&quot;C&quot;</span>,<span class="st">&quot;T&quot;</span>,<span class="st">&quot;TA.T&quot;</span>)
<span class="co">#&gt; [1] &quot;AID&quot;          &quot;AID(hotspot)&quot;</span></code></pre></div>
<p>As visible in the default table, it is possible to use alternative nucleotide symbols. However it is not possible to use these symbols in the look up mutation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cMut<span class="op">::</span><span class="kw">linkPatterns</span>(<span class="st">&quot;C&quot;</span>,<span class="st">&quot;D&quot;</span>,<span class="st">&quot;A.A&quot;</span>)
<span class="co">#&gt; Error in cMut::linkPatterns(&quot;C&quot;, &quot;D&quot;, &quot;A.A&quot;): !grepl(&quot;[^ACGTacgt]&quot;, alt) is not TRUE</span></code></pre></div>
</div>
<div id="groupclusters" class="section level2">
<h2>groupClusters</h2>
<p>The <code>cMut::groupClusters()</code> function is best used as an optional followup function for <code>cMut::identifyAndAnnotateClusters()</code>. This function will group the data based on the cluster IDs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">groups &lt;-<span class="st"> </span>cMut<span class="op">::</span><span class="kw">groupClusters</span>(<span class="dt">table =</span> results,
                    <span class="dt">tibble =</span> <span class="ot">FALSE</span>)
<span class="co"># The cMuts column contains lists of mutations that belong to the cluster.</span>
<span class="co"># It's excluded because it messed up the format.</span>
groups[,<span class="op">-</span><span class="kw">which</span>(<span class="kw">names</span>(groups) <span class="op">==</span><span class="st"> &quot;cMuts&quot;</span>)] 
<span class="co">#&gt;             clusterId refs alts plusStrand minusStrand       clusterType</span>
<span class="co">#&gt; 1  chr10 testSample 6 T, G C, A    T&gt;C G&gt;A     C&gt;T A&gt;G C&gt;T A&gt;G / T&gt;C G&gt;A</span>
<span class="co">#&gt; 2 chr17 testSample 12 C, G A, C    C&gt;A G&gt;C     C&gt;G G&gt;T C&gt;A G&gt;C / C&gt;G G&gt;T</span>
<span class="co">#&gt; 3  chr2 testSample 27 A, T G, C    A&gt;G T&gt;C     A&gt;G T&gt;C A&gt;G T&gt;C / A&gt;G T&gt;C</span>
<span class="co">#&gt; 4  chr5 testSample 55 T, T A, A    T&gt;A T&gt;A     A&gt;T A&gt;T A&gt;T A&gt;T / T&gt;A T&gt;A</span>
<span class="co">#&gt; 5  chr7 testSample 13 G, G T, T    G&gt;T G&gt;T     C&gt;A C&gt;A C&gt;A C&gt;A / G&gt;T G&gt;T</span>
<span class="co">#&gt; 6  chr9 testSample 48 G, A A, T    G&gt;A A&gt;T     T&gt;A C&gt;T G&gt;A A&gt;T / T&gt;A C&gt;T</span></code></pre></div>
<p>If the <code>linkPatterns = TRUE</code> parameter is used during the <code>cMut::identifyAndAnnotateClusters()</code> function or the custom table have a similar column then it is possible to use the <code>patternIntersect = TRUE</code> parameter in the <code>cMut::groupClusters()</code> to add a column with the intersection of the patterns per cluster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">groups &lt;-<span class="st"> </span>cMut<span class="op">::</span><span class="kw">groupClusters</span>(<span class="dt">table =</span> results,
                    <span class="dt">patternIntersect =</span> <span class="ot">TRUE</span>)
<span class="co"># The cMuts column contains lists of mutations that belong to the cluster.</span>
<span class="co"># It's excluded because it messed up the format.</span>
groups[,<span class="op">-</span><span class="kw">which</span>(<span class="kw">names</span>(groups) <span class="op">==</span><span class="st"> &quot;cMuts&quot;</span>)] 
<span class="co">#&gt; # A tibble: 6 x 8</span>
<span class="co">#&gt;   clusterId refs  alts  plusStrand minusStrand clusterType patternIntersect</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;lis&gt; &lt;lis&gt; &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt;       &lt;list&gt;          </span>
<span class="co">#&gt; 1 chr10 te~ &lt;chr~ &lt;chr~ T&gt;C G&gt;A    C&gt;T A&gt;G     C&gt;T A&gt;G / ~ &lt;list [0]&gt;      </span>
<span class="co">#&gt; 2 chr17 te~ &lt;chr~ &lt;chr~ C&gt;A G&gt;C    C&gt;G G&gt;T     C&gt;A G&gt;C / ~ &lt;list [0]&gt;      </span>
<span class="co">#&gt; 3 chr2 tes~ &lt;chr~ &lt;chr~ A&gt;G T&gt;C    A&gt;G T&gt;C     A&gt;G T&gt;C / ~ &lt;list [1]&gt;      </span>
<span class="co">#&gt; 4 chr5 tes~ &lt;chr~ &lt;chr~ T&gt;A T&gt;A    A&gt;T A&gt;T     A&gt;T A&gt;T / ~ &lt;list [0]&gt;      </span>
<span class="co">#&gt; 5 chr7 tes~ &lt;chr~ &lt;chr~ G&gt;T G&gt;T    C&gt;A C&gt;A     C&gt;A C&gt;A / ~ &lt;list [0]&gt;      </span>
<span class="co">#&gt; 6 chr9 tes~ &lt;chr~ &lt;chr~ G&gt;A A&gt;T    T&gt;A C&gt;T     G&gt;A A&gt;T / ~ &lt;list [0]&gt;      </span>
<span class="co">#&gt; # ... with 1 more variable: has.intersect &lt;lgl&gt;</span></code></pre></div>
<div id="bugs-questions-suggestions" class="section level3">
<h3>Bugs, questions, suggestions?</h3>
<p>Feel free to the contact the creator! (^__^) See info in the DESCRIPTION file.</p>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
